### PR TITLE
Support trilinear sampling in Resize CPU and CUDA kernels

### PR DESCRIPTION
**Pull request by hariharans29**
**Description**: Support 3-D input (or 5-D input with outermost 2 scales being 1.0 - batched multi-channel volumes [N,C, D, H, W]) for the "Linear" mode in the CPU and CUDA Resize kernels

**Motivation and Context**
Support a first party ONNX model in ORT

### CODE REVIEW

File: onnxruntime/core/providers/cpu/tensor/upsample.cc [hunk 1]
```diff
@@ -363,8 +385,8 @@ void UpsampleBilinear(int64_t batch_size,
     in_x1[x] = std::min(static_cast<int64_t>(in_x), input_width - 1);
     in_x2[x] = std::min(in_x1[x] + 1, input_width - 1);
 
-    dx1[x] = std::abs(in_x - in_x1[x]);
-    dx2[x] = std::abs(in_x - in_x2[x]);
+    dx1[x] = std::fabs(in_x - in_x1[x]);
```
Comment (hariharans29) [inline onnxruntime/core/providers/cpu/tensor/upsample.cc:388]: For consistency's sake - we use fabs everywhere else in this file

File: onnxruntime/test/providers/cpu/tensor/resize_op_test.cc [hunk 2]
```diff
@@ -197,7 +197,66 @@ TEST(ResizeOpTest, ResizeOpLineartUpSampleTest_2DBilinear_align_corners) {
   test.Run();
 }
 
-TEST(ResizeOpTest, ResizeOpLineartScalesNoOpTest) {
+TEST(ResizeOpTest, ResizeOpLinearDownSampleTest_3DTrilinear_pytorch_half_pixel) {
```
Comment (hariharans29) [inline onnxruntime/test/providers/cpu/tensor/resize_op_test.cc:200]: 3D + downsample

File: onnxruntime/test/providers/cpu/tensor/resize_op_test.cc [hunk 3]
```diff
@@ -197,7 +197,66 @@ TEST(ResizeOpTest, ResizeOpLineartUpSampleTest_2DBilinear_align_corners) {
...
+  test.AddInput<float>("scales", {0}, scales);
+  test.AddInput<int64_t>("sizes", {3}, sizes);
+
+  std::vector<float> Y = {1.6666666f, 7.0f, 12.333333f};
+
+  test.AddOutput<float>("Y", {sizes[0], sizes[1], sizes[2]}, Y);
+  test.Run(OpTester::ExpectResult::kExpectSuccess, "", {kTensorrtExecutionProvider});  // TensorRT: results mismatch
+}
+
+TEST(ResizeOpTest, ResizeOpLinearUpSampleTest_5DTrilinear_pytorch_half_pixel) {
```
Comment (hariharans29) [inline onnxruntime/test/providers/cpu/tensor/resize_op_test.cc:233]: 5D input + upsample

File: onnxruntime/core/providers/cuda/tensor/resize_impl.h [hunk 4]
```diff
@@ -23,7 +23,7 @@ void ResizeImpl(
     TArray<int64_t>& input_strides,
     TArray<fast_divmod>& output_div_pitches,
     TArray<float>& scales_vals,
-    TArray<float>& roi,
+    TArray<float, 10>& roi,
```
Comment (hariharans29) [inline onnxruntime/core/providers/cuda/tensor/resize_impl.h:26]: default capacity is 8 which doesn't compile when input is 5D and roi vector contains 10 elements

File: onnxruntime/core/providers/cpu/tensor/upsample.cc [hunk 5]
```diff
@@ -310,24 +310,46 @@ void UpsampleBilinear(int64_t batch_size,
                       AllocatorPtr& alloc,
                       GetOriginalCoordinateFunc get_original_coordinate) {
   std::vector<float> y_original;
+  y_original.reserve(output_height);
```
Comment (hariharans29) [inline onnxruntime/core/providers/cpu/tensor/upsample.cc:313]: best to reserve memory - we know the number of elements that we are going to put into this container.

**Comment by askhade** [comment]
Can you validate these newly supported scenarios using onnx's implementation of resize. Basically use the test data from your test cases with the resize implementation here https://github.com/onnx/onnx/blob/master/onnx/backend/test/case/node/resize.py and see whether the output's match.
This will help to validate whether the implementation in ort matches the spec.

**Comment by hariharans29** [comment]
> Can you validate these newly supported scenarios using onnx's implementation of resize. Basically use the test data from your test cases with the resize implementation here https://github.com/onnx/onnx/blob/master/onnx/backend/test/case/node/resize.py and see whether the output's match.
> This will help to validate whether the implementation in ort matches the spec.

I believe I did that for the upsampling test. Let me run the downsampling test data foe completeness’ sake.

**Review by zhanghuanrong — APPROVED** [review]

**Review by askhade — APPROVED** [review]

**Pull request merged**